### DISCUSSION TITLE
QOL feature: Fill in image size

### DISCUSSION THREAD

**Discussion by Woisek** [vladmandic/sdnext#325] [Ideas]
When I'm doing upscales in img2img, I mostly drag and drop the images into the image field. But when doing so, the dimension of the image is not reflected/updated in the width and hight inputs. And instead of fiddeling and look up the image size, it would be very convienient if we had a button that fills in the correct size of the current image into the corresponding fields. Would be a handy time saver. 😋

**Comment by gruevy** [comment]
All you have to do, if the .png file has the generation info, is drop it into the prompt field and hit this button. Works on several different tabs.
![image](https://github.com/vladmandic/automatic/assets/1585698/415d5275-1476-4328-b0b4-a9f2a8dac525)

**Reply by Woisek** [reply]
I know this feature since the latest updates. But this isn't what I have requested, please read my feature wish again to understand.

**Comment by vladmandic** [comment]
first, check this is selected: UI -> Settings -> User Interface -> Send size when sending prompt or image to another interface
second, if you drop an image into process image tab and then use send to it will populate size.
doing that directly in txt2img or img2img can be dangerous, I'd prefer not to do it. but if someone submits a PR, I'd take a look at it.

**Reply by Woisek** [reply]
I know about the "send image" function. But why do I have to do this additional step and having reset (all) my options on the target page, just for getting the size? When I do multiple images for specific resizing, where nothing is changed besides the upscale value, I am at the img2img page and just drag and drop images one after another into the image holder and press "generate".

And please, what exactly is dangerous to a button, that executes a script part that reads out and sets the image size value of the currently used image? Basically the same script part, that is used when the "send image to" function is running ... ?

Plus, if you are not aware, this "dangerous" practice is done already as an extension.
https://github.com/alemelis/sd-webui-ar.git gives us buttons, that sets the width and height of an image we want to to create to a defined value. An incredible helpful extension.
But what I need in this case is not a defined value, but the actual size value of an existing and used image.

**Reply by vladmandic** [reply]
its dangerous from user perspective and dev perspective - it would be very easy to load 4k image which sets resolution to something that cannot work in txt2img or img2img. going via process image is at least more explicit.
from dev perspective, all the transfer is done via triggering copy&paste buttons and that is the ugliest code on the planet. chances of creating new issues and/or incompatibilities with extensions are beyond reasonable - reward is not worth the risk. but if someone wants to contribute a clean and well-tested pr, by all means - go for it.

**Reply by Woisek** [reply]
>it would be very easy to load 4k image which sets resolution to something that cannot work in txt2img or img2img.

User generate images at 6k, 8k and even 16k (https://www.reddit.com/r/StableDiffusion/comments/13bc8qj/just_learned_how_to_use_controlnet_11_tile_with/), so ...

>from dev perspective, all the transfer is done via triggering copy&paste buttons

What transfer you are talking about ... ?

>creating new issues and/or incompatibilities with extensions are beyond reasonable

Oh boy ... we are talking about getting the size of an image and paste those values into two text fields, surely a highly ultra complex process, effecting the whole eco system ... 😦
Anyway ...

**Reply by vladmandic** [reply]
No need for sarcasm. Yes, it's hugely complex if you knew how that code worked and it's one of my least favorite part of the code. Feel free to submit a PR.

**Reply by Aptronymist** [reply]
> > it would be very easy to load 4k image which sets resolution to something that cannot work in txt2img or img2img.
> 
> User generate images at 6k, 8k and even 16k ([reddit.com/r/StableDiffusion/comments/13bc8qj/just_learned_how_to_use_controlnet_11_tile_with](https://www.reddit.com/r/StableDiffusion/comments/13bc8qj/just_learned_how_to_use_controlnet_11_tile_with/)), so ...
> 
> > from dev perspective, all the transfer is done via triggering copy&paste buttons
> 
> What transfer you are talking about ... ?
> 
> > creating new issues and/or incompatibilities with extensions are beyond reasonable
> 
> Oh boy ... we are talking about getting the size of an image and paste those values into two text fields, surely a highly ultra complex process, effecting the whole eco system ... 😦 Anyway ...

I believe what Vlad is nicely trying to tell you is that it's a risk in the sense of potentially crashing the application entirely if the image size was too large or perhaps had bad metadata. 
It might even create an exploitable security vulnerability if malformed images were imported. Things of that nature. 
It's far more complex than you make it out to be, and he has to take that into account, because no doubt thousands of people are using his code (including myself), whereas you do not, and are not responsible or accountable for anything.

He's been doing a stunningly great job so far, I trust his instincts on this, you probably should too, so don't be rude to him because he didn't tell you what you wanted to hear.

**Reply by Woisek** [reply]
>I believe what Vlad is nicely trying to tell you is that it's a risk in the sense of potentially crashing the application entirely if the image size was too large or perhaps had bad metadata.

Again, people generate images at 16k which sure will not be the standard but an exception and it's obviously possible. So, what size is "too large" in your opinion? And I don't talk about meta data, I'm talking of getting the image size and populate 2 text field with it.

>It might even create an exploitable security vulnerability if malformed images were imported. Things of that nature.

I'm a developer myself. I'm coding websites where people can edit their content via my own developed CMS. For over the last 15 years since I'm doing this, I had _never_ any case of "an exploitable security vulnerability of malformed image". The chance of this happening will probably tend right to zero.
That said, please enlighten me from where this malformed images come from, that a user would put into WebUI. Also, a "malformed image" would be possible deleted once a user noticed it, because it probably won't open in his standard image viewer. And I hope you won't tell me that the user then goes: "Oh, I can't see this in my image viewer, so I throw it into WebUI ..."

>It's far more complex than you make it out to be, 

Again, I see no complexity in reading out an images dimension and setting two values into text fields. I must have done something wrong my whole time as developer.

>and he has to take that into account, because no doubt thousands of people are using his code (including myself), whereas you do not, and are not responsible or accountable for anything.

Right, because it's not "his" project anymore since release, more needs and option wishes come up, naturally. And as I stated multiple times, I'm just trying to help, because I like and believe in this project. And as a developer I know exactly how it feels to be pushed by outsiders. That's why I try to only make comments and ideas where I think it makes sense. not only for me, but also with regard to all other users. I've already told vlad this so that he doesn't think I'm just trying to be annoying here. That's far from my mind.

Personally, as a web developer, I am confronted again and again with "ideas" from users, which are indeed mostly short-sighted and/or ill-considered, simply because they don't understand the works behind it.
Therefore, my intention is not to be such a user, but to provide well-considered ideas that are on par with the developer. Which is logical, because I want to be able to work with the thing in a meaningful way.

But that doesn't mean I'm not open to counter-arguments. I'd just like them to be well-founded, so that I, as the developer, not a user, can understand them.
And if you say that you would like to have two values read from an image and inserted into text fields, and the answer is that this is highly complex and extremely dangerous, even though this function is already part of the processing ... I'm sorry, this is simply not comprehensible and just ridiculous.
I would seriously prefer the answer: "No, I don't want to do it" than to be presented with something like this. That would at least be more honest.

>He's been doing a stunningly great job so far, I trust his instincts on this, you probably should too, so don't be rude to him because he didn't tell you what you wanted to hear.

I think that vlad does an extremely extraordinary job, which is the reason why I try to get involved here so much. I've told him that before and I'll say it again: @vladmandic , you are doing an extremely great job here, your effort is much appreciated.
I don't think I would have been rude. If it came across that way, then I'm sorry, because of course that was never the intention at any time.

And of course the developer knows best how his project has to be handled. Nevertheless, it is important to allow feedback from outside, because there is always a project blindness that hits every developer, myself included. If you're too deep into the matter, it's easy to overlook things that are simple or not so good.
Personally, I would like to have someone who is both a user and a developer to be able to draw my attention to things that I overlooked in my blindness or that I didn't even think of during development because I work completely differently to other users.

So, if vlad tells me straight: "Hey, I don't need your opinion, I'll do what I think is right", then I will respect that and stop interfering. Because the last thing I want is to waste my own time trying to annoy someone who has important things to do.

**Comment by vladmandic** [comment]
> Again, I see no complexity in reading out an images dimension and setting two values into text fields. I must have done something wrong my whole time as developer.

Because that's unfortunately NOT **how this part of the code works**.  
I said several times this is probably my least favorite part of the code as it uses multiple self-registered param list and passing params is a blind list with specific length that gets passed to multiple tabs via anonymous gradio updates. so checking list length and parsing it correctly is a nightmare - different parts of the code expect list of specific length and there is no param names passed. There is NO function that that "sets values in text fields" as such. So its the question of risk vs reward - this is high risk request vs small reward.

> So, if vlad tells me straight: "Hey, I don't need your opinion, I'll do what I think is right", then I will respect that and stop interfering. Because the last thing I want is to waste my own time trying to annoy someone who has important things to do.

I think I demonstrated plenty of times that I'm open for opinions. And even changed mine multiple times in the past based on user feedback.

I personally think this is not worth it, but I also said multiple times that I'd welcome a PR. So if you or someone actually writes and tests code, I'd be happy to review it and merge it.
